### PR TITLE
Add pause&resume block job test case for drive_mirror.

### PR DESCRIPTION
**Pull request by vivianQizhu**
Add pause&resume block job test case for drive_mirror.
Signed-off-by: Qianqian Zhu qizhu@redhat.com
id: 1360154

### CODE REVIEW

**Comment by vivianQizhu** [comment]
Depends on: https://github.com/avocado-framework/avocado-vt/pull/637

File: qemu/tests/block_copy.py [hunk 1]
```diff
@@ -136,6 +136,40 @@ def is_cancelled():
...
 
+    def is_paused(self):
+        """
+        Return block job paused status.
+        """
+        paused = self.get_status().get("paused")
+        offset_p = self.get_status().get("offset")
+        time.sleep(random.uniform(1, 3))
+        offset_l = self.get_status().get("offset")
+        paused &= bool(offset_p == offset_l)
```
Comment (xutian) [inline qemu/tests/block_copy.py]: No need to convert here.

File: qemu/tests/block_copy.py [hunk 2]
```diff
@@ -136,6 +136,40 @@ def is_cancelled():
...
+        return paused
+
+    def pause_job(self):
+        """
+        pause active job;
+        """
+        if self.is_paused():
```
Comment (xutian) [inline qemu/tests/block_copy.py]: pause a paused job is not allowed by qemu, so here no need to check it.

**Comment by xutian** [comment]
we also need to  check qmp event  after block job resumed/paused.

**Comment by vivianQizhu** [comment]
@xutian There are no such events for block job resumed/paused. Thanks.

File: qemu/tests/block_copy.py (hunk 2 above)
Comment (vivianQizhu) [inline qemu/tests/block_copy.py]: @xutian I've just tested, qemu won't prevent this action, so I think it is necessary to keep this check. Thanks.

Comment (xutian) [inline qemu/tests/block_copy.py]: Please check with libvirt, if still not, please fill a bug to qemu and libvirt

**Comment by xutian** [comment]
I noticed it, just have a try with qemu-2.6, really no any event after pause/resume block job. can you ask developer why no such kind of event raise after pause/resume block jobs.

**Comment by xutian** [comment]
{"QMP": {"version": {"qemu": {"micro": 0, "minor": 6, "major": 2}, "package": " (qemu-kvm-rhev-2.6.0-5.el7)"}, "capabilities": []}}
{"execute": "qmp-capabilities"}
{"return": {}}
{"execute": "drive-mirror", "arguments": {"device":"d1", "target":"/tmp/aa.qcow2", "sync": "full"}}
{"return": {}}
{"execute": "block-job-pause", "arguments": {"device":"d1"}}
{"return": {}}
{"execute": "block-job-resume", "arguments": {"device":"d1"}}
{"return": {}}
{"timestamp": {"seconds": 1469590646, "microseconds": 18940}, "event": "BLOCK_JOB_READY", "data": {"device": "d1", "len": 7006388224, "offset": 7006388224, "speed": 0, "type": "mirror"}}

File: qemu/tests/block_copy.py (hunk 2 above)
Comment (vivianQizhu) [inline qemu/tests/block_copy.py]: @xutian To pause a paused job wont cause any error for qemu, so I think there is no any bug for this funciton for qemu. Keep this check just for informing tester job already paused.

**Comment by vivianQizhu** [comment]
@xutian I don't think such event is a must have function, it is acceptable for me.

File: qemu/tests/block_copy.py [hunk 3]
```diff
@@ -136,6 +136,40 @@ def is_cancelled():
...
 
+    def is_paused(self):
+        """
+        Return block job paused status.
+        """
+        paused = self.get_status().get("paused")
+        offset_p = self.get_status().get("offset")
+        time.sleep(random.uniform(1, 3))
+        offset_l = self.get_status().get("offset")
+        paused &= offset_p == offset_l
```
Comment (xutian) [inline qemu/tests/block_copy.py]: paused &= (offset_p == offset_l)

Comment (vivianQizhu) [inline qemu/tests/block_copy.py]: @xutian , operator '==' has higher priority than &, so we should not need () here.  I've tested it, it works correctly.

**Comment by vivianQizhu** [comment]
16:05:16 DEBUG| Send command: {'execute': 'drive-mirror', 'arguments': {'mode': 'absolute-paths', 'format': 'qcow2', 'device': u'drive_image1', 'speed': 0, 'sync': 'full', 'target': '/usr/share/avocado/data/avocado-vt/images/target1.qcow2'}, 'id': 'fdQLHXXn'}
16:05:19 INFO | Pause block job.
16:05:19 DEBUG| (monitor qmp1) Sending command 'block-job-pause'
16:05:26 INFO | Resume block job.
16:05:26 DEBUG| (monitor qmp1) Sending command 'block-job-resume' 
16:07:25 INFO | PASS 1-smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.my-provider.drive_mirror.simple_test.pause_resume_job.2qcow2

**Comment by xutian** [comment]
Acked,  thanks @vivianQizhu

**Comment by vivianQizhu** [comment]
@CongLi , Would you please help review this? Thanks.

**Comment by CongLi** [comment]
Acked-by: Cong Li <coli@redhat.com>

**Comment by CongLi** [comment]
Thanks @vivianQizhu @xutian , merged.

File: qemu/tests/block_copy.py (hunk 3 above)
Comment (yanan-fu) [inline qemu/tests/block_copy.py]: This is not work with rhel6 host.  As in rhel6 host, "query-block-jobs" has no option "paused", so in the code,
the variable "paused" is NoneType,  and  "paused &= offset_p == offset_1" will failed as "TypeError: unsupported operand type(s) for &=: 'NoneType' and 'bool'" 

rhel 6.8:
{"execute": "query-block-jobs", "id": "vix3JVGK"}
{"return": [{"device": "drive_image1", "len": 21474836480, "offset": 14697, "speed": 0, "type": "mirror"}], "id": "vix3JVGK"}

rhel 7.3:
{"execute": "query-block-jobs", "id": "vmaMoZC0"}
{"return": [{"io-status": "ok", "device": "drive_image1", "busy": true, "len": 3719692288, "offset": 52494336, "paused": false, "speed": 0, "ready": false, "type": "mirror"}], "id": "vmaMoZC0"}

**Pull request merged**